### PR TITLE
Nexus: support supercell twists in PySCF workflows

### PR DESCRIPTION
**Pull request by jtkrogel**
## Proposed changes

This PR adds support for arbitrary supercell twists/twist grids in workflows involving PySCF and QMCPACK.

This PR is now ready for review.

## What type(s) of changes does this code introduce?

- New feature

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?

Laptop, Improv at ALCF

## Checklist

- Yes. This PR is up to date with current the current state of 'develop'

## Path out of WIP

- [x] Supercell twist to primitive cell k-point mappings in generated PySCF inputs
- [x] Loop-free twist averaging in subsequent QMCPACK runs
- [x] Add a twist-averaging VMC examples for diamond
- [x] Fix pyscf tests

### CODE REVIEW

**Comment by jtkrogel** [comment]
Ready to proceed.

**Comment by ye-luo** [comment]
could you suggest a reviewer?

**Comment by jtkrogel** [comment]
Yes, already flagged @anbenali for this.

**Comment by ye-luo** [comment]
> Yes, already flagged @anbenali for this.

Oh thanks. I missed it.

**Review by prckent — COMMENTED** [review]
Noting that I am currently running the examples locally. These appear to be running OK. PySCF convergence tolerance is high so they take some minutes.

ntest_nexus_pyscf_input failed on me, but I have yet to investigate further
```
1452: Test name     : pyscf_input
1452: Test sublabel : test_write
1452: Test exception: "AssertionError: "
1452: Test backtrace:
1452:   File "/home/pk7/projects/qmc/git_QMCPACK_prckent/qmcpack/nexus/bin/nxs-test", line 478, in run
1452:     self.operation()
1452:   File "/home/pk7/projects/qmc/git_QMCPACK_prckent/qmcpack/nexus/bin/nxs-test", line 1210, in pyscf_input
1452:     nunit('write')
1452:   File "/home/pk7/projects/qmc/git_QMCPACK_prckent/qmcpack/nexus/bin/nxs-test", line 349, in nunit
1452:     run_external_unit_test(test_name,unit_test)
1452:   File "/home/pk7/projects/qmc/git_QMCPACK_prckent/qmcpack/nexus/bin/nxs-test", line 388, in run_external_unit_test
1452:     unit_test()
1452:   File "/home/pk7/projects/qmc/git_QMCPACK_prckent/qmcpack/nexus/tests/unit/test_pyscf_input.py", line 577, in test_write
1452:     assert(text_eq(text,ref_text))
1452:            ^^^^^^^^^^^^^^^^^^^^^^
1452:
1452: Test status: fail
```
Hopefully unrelated, but note that my installation of graphviz can't produce PNGs, so this trips up a few tests all the time.

**Comment by jtkrogel** [comment]
For the nexus test failure, please add two print statements after this line to investigate:

```
1452:   File "/home/pk7/projects/qmc/git_QMCPACK_prckent/qmcpack/nexus/tests/unit/test_pyscf_input.py", line 577, in test_write
1452:     assert(text_eq(text,ref_text))

# add these
print(ref_text)
print(text)
```

For the examples added with this PR, `diamond_pp_hf_twistavg_prim.py` (primitive cell twist averaging) should run cleanly (at least it did for me -- please post the converter output), while `diamond_pp_hf_twistavg.py` (supercell twist averaging) currently fails at the converter level due to changes needed to Anouar's `savetoqmcpack`.  

These features have bug-fixes needed at the QMCPACK/QMCPACK-converter levels (I've made Anouar aware already).  This PR implements the Nexus-side features needed to drive these workflows, but does not guarantee that QMCPACK and its converters function properly.

**Comment by prckent** [comment]
Thanks Jaron -- the situation is clear to me now. @anbenali How far off are the updates to savetoqmcpack? I put PySCF 2.6.2 in spack so can easily test the latest version. In would be nice to have patch so that Nexus support can be tested, but perhaps there are puzzles to solve?

**Comment by prckent** [comment]
At least one difference is that one entry in newly generated sp_kpoints array is not quite zero:
```
sp_kpoints = array( [ 
2597:      [ -2.9375236054174356e-17, 0.0, 0.0 ]  ] )
2597:
```

Selected output from `ctest -R ntest_nexus_pyscf_input`
```
597: Test timeout computed to be: 3600
2597: 
2597:         #! /usr/bin/env python3
2597:         
2597:         from pyscf import scf
2597:         
2597:         
2597:         ### generated system text ###
2597:         from numpy import array
2597:         from pyscf.pbc import gto as gto_loc
2597:         cell = gto_loc.Cell()
2597:         cell.a             = '''
2597:                              1.78500000   1.78500000   0.00000000
2597:                              0.00000000   1.78500000   1.78500000
2597:                              1.78500000   0.00000000   1.78500000
2597:                              '''
2597:         cell.basis         = 'bfd-vdz'
2597:         cell.dimension     = 3
2597:         cell.ecp           = 'bfd'
2597:         cell.unit          = 'A'
2597:         cell.atom          = '''
2597:                              C    0.00000000   0.00000000   0.00000000
2597:                              C    0.89250000   0.89250000   0.89250000
2597:                              '''
2597:         cell.drop_exponent = 0.1
2597:         cell.verbose       = 5
2597:         cell.charge        = 0
2597:         cell.spin          = 0
2597:         cell.build()
2597:         kpts = array( [ 
2597:              [ 0.0, 0.0, 0.0 ]  ,
2597:              [ 0.4656748546088228, 0.4656748546088228, -0.4656748546088228 ]  ] )
2597:         ### end generated system text ###
2597:         
2597:         
2597:         
2597:         mf = scf.RHF(mol)
2597:         mf.kernel()
2597:         
2597:         ### generated conversion text ###
2597:         from PyscfToQmcpack import savetoqmcpack
2597:         tiling =  [ 2,1,1 ] 
2597:         sp_kpoints = array( [ 
2597:              [ 0.0, 0.0, 0.0 ]  ] )
2597:         sp_kmap    = array( [ 
2597:              [ 0, 1 ]  ] )
2597:         for n,kp in enumerate(sp_kpoints):
2597:             savetoqmcpack(cell,mf,'scf.twistnum_{}'.format(str(n).zfill(3)),kmesh=tiling,kpts=kpts [ sp_kmap [ n ]  ] ,sp_twist=kp)
2597:         #end for
2597:         ### end generated conversion text ###
2597:         
2597: #! /usr/bin/env python3
2597: 
2597: from pyscf import scf
2597: 
2597: 
2597: ### generated system text ###
2597: from numpy import array
2597: from pyscf.pbc import gto as gto_loc
2597: cell = gto_loc.Cell()
2597: cell.a             = '''
2597:                      1.78500000   1.78500000   0.00000000
2597:                      0.00000000   1.78500000   1.78500000
2597:                      1.78500000   0.00000000   1.78500000
2597:                      '''
2597: cell.basis         = 'bfd-vdz'
2597: cell.dimension     = 3
2597: cell.ecp           = 'bfd'
2597: cell.unit          = 'A'
2597: cell.atom          = '''
2597:                      C    0.00000000   0.00000000   0.00000000
2597:                      C    0.89250000   0.89250000   0.89250000
2597:                      '''
2597: cell.drop_exponent = 0.1
2597: cell.verbose       = 5
2597: cell.charge        = 0
2597: cell.spin          = 0
2597: cell.build()
2597: kpts = array( [ 
2597:      [ 0.0, 0.0, 0.0 ] ,
2597:      [ 0.4656748546088228, 0.4656748546088228, -0.4656748546088228 ]  ] )
2597: ### end generated system text ###
2597: 
2597: 
2597: 
2597: mf = scf.RHF(mol)
2597: mf.kernel()
2597: 
2597: ### generated conversion text ###
2597: from PyscfToQmcpack import savetoqmcpack
2597: tiling =  [ 2,1,1 ] 
2597: sp_kpoints = array( [ 
2597:      [ -2.9375236054174356e-17, 0.0, 0.0 ]  ] )
2597: sp_kmap    = array( [ 
2597:      [ 0, 1 ]  ] )
2597: for n,kp in enumerate(sp_kpoints):
2597:     savetoqmcpack(cell,mf,'scf.twistnum_{}'.format(str(n).zfill(3)),kmesh=tiling,kpts=kpts [ sp_kmap [ n ]  ] ,sp_twist=kp)
2597: #end for
2597: ### end generated conversion text ###
2597: 
2597: 
2597: Test name     : pyscf_input
2597: Test sublabel : test_write
2597: Test exception: "AssertionError: "
2597: Test backtrace:
2597:   File "/home/pk7/projects/qmc/git_QMCPACK_prckent/qmcpack/nexus/bin/nxs-test", line 478, in run
2597:     self.operation()
2597:   File "/home/pk7/projects/qmc/git_QMCPACK_prckent/qmcpack/nexus/bin/nxs-test", line 1210, in pyscf_input
2597:     nunit('write')
2597:   File "/home/pk7/projects/qmc/git_QMCPACK_prckent/qmcpack/nexus/bin/nxs-test", line 349, in nunit
2597:     run_external_unit_test(test_name,unit_test)
2597:   File "/home/pk7/projects/qmc/git_QMCPACK_prckent/qmcpack/nexus/bin/nxs-test", line 388, in run_external_unit_test
2597:     unit_test()
2597:   File "/home/pk7/projects/qmc/git_QMCPACK_prckent/qmcpack/nexus/tests/unit/test_pyscf_input.py", line 580, in test_write
2597:     assert(text_eq(text,ref_text))
2597:            ^^^^^^^^^^^^^^^^^^^^^^
2597: 
2597: Test status: fail
1/1 Test #2597: ntest_nexus_pyscf_input ..........***Failed    0.63 sec

0% tests passed, 1 tests failed out of 1
```

via
```
diff --git a/nexus/tests/unit/test_pyscf_input.py b/nexus/tests/unit/test_pyscf_input.py
index a966746cc..951435e01 100644
--- a/nexus/tests/unit/test_pyscf_input.py
+++ b/nexus/tests/unit/test_pyscf_input.py
@@ -574,6 +574,9 @@ def test_write():
     text = text.replace('[',' [ ').replace(']',' ] ')
     ref_text = ref_text.replace('[',' [ ').replace(']',' ] ')
 
+# add these
+    print(ref_text)
+    print(text)
     assert(text_eq(text,ref_text))
 
 #end def test_write
```

Note that this is failing in CI as well -- the centos builds show the same issue. Presumably either something is not zeroed  out or rounding is slightly different and a numerical tolerance is needed.

File: nexus/tests/unit/test_pyscf_input.py [hunk 1]
```diff
@@ -554,10 +554,23 @@ def test_write():
...
         from PyscfToQmcpack import savetoqmcpack
-        savetoqmcpack(cell,mf,'scf',kpts)
+        tiling = [2,1,1]
+        sp_kpoints = array([
+            [0.0, 0.0, 0.0]])
+        sp_kmap    = array([
+            [0, 1]])
+        for n,kp in enumerate(sp_kpoints):
+            savetoqmcpack(cell,mf,'scf.twistnum_{{}}'.format(str(n).zfill(3)),kmesh=tiling,kpts=kpts[sp_kmap[n]],sp_twist=kp)
+        #end for
```
Comment (anbenali) [inline nexus/tests/unit/test_pyscf_input.py:564]: In order for this to work with kpoints in a ncsf style, we need to also pass the sp_kmap. The reason is that when running N kpts for example, mf method has N mo_energies and N mo_coeffs. but the tiling will only set a subset of kpts. we need then to use the index of the kptks provided by the sp_kmap. 

 savetoqmcpack(cell,mf,'scf.twistnum_{{}}'.format(str(n).zfill(3)),kmesh=tiling,kpts=kpts[sp_kmap[n]],sp_twist=kp,kmap=sp_kmap[n])

**Comment by anbenali** [comment]
Everything is working on my end. I fixed the files that did not work. Please check if I forgot something.

**Comment by prckent** [comment]
Test this please

**Comment by prckent** [comment]
^^^ To get some more variety in the test configurations run.  Hopefully only the floating point rounding/tolerance issue needs to be tackled.

Thanks Anouar. Great to have this path supported.

**Review by prckent — CHANGES_REQUESTED** [review]
(Noting for ease of seeing the status on the PR summary)

The k-points tolerance issue should be addressed - besides CentOS failures, presumably this is what is causing the failures on sulfur (Intel RHEL9), since the same test is failing.

Definitely not needed for this PR, but from the point of view of reading CI failures, it would be very helpful to replace all instances of assert(text_eq(a,b)) with a custom function that gave actionable info by printing the failure and the a,b before aborting.

**Comment by anbenali** [comment]
While driving. I realized that the QMC input requires explicitly the twist
coordinate to compute the phase to apply to the orbital.  In the current
qmc input from nexus we get a twist number. I am not sure if that is
accessed in the H5 file or of of it os even stored. I can look at it in the
afternoon.

**Comment by prckent** [comment]
@anbenali There is an indentation bug (or more) in PyscfToQmcpack.py. I got the following running the diamond prim example:
```
Traceback (most recent call last):
  File "/home/pk7/projects/qmc/git_QMCPACK_prckent/qmcpack/nexus/examples/qmcpack/rsqmc_pyscf/02_diamond_hf_qmc/runs/diamond_ta_prim/scf/scf.py", line 45, in <module>
    from PyscfToQmcpack import savetoqmcpack
  File "/home/pk7/projects/qmc/git_QMCPACK_prckent/qmcpack/src/QMCTools/PyscfToQmcpack.py", line 553
    E_g, C_gamma,E_g_unsorted,C_unsorted = mo_k2gamma(cell, e_k, mo_k, kpts,kmesh)
                                                                                  ^
IndentationError: unindent does not match any outer indentation level
```

**Comment by prckent** [comment]
Interestingly what happens in that example is that on my system the pyscf output step fails as above, then the converter step runs and also fails, since the HDF5 files is not present. Somehow though, Nexus does catch the failures and the converter step runs forever. Maybe something in the workstation class (?)

**Comment by anbenali** [comment]
@jtkrogel: there are still errors to be fixed for the kpts case.
At this point you can try changing  "drop_exponent = 0.3" so it would run faster for the tests. 

Then if you have 2 twists generated, we pass the following "title='scf.twistnum_{}'.format(str(n).zfill(3))' 
but then we do a static link to scf.h5 for c4q.orbs.h5. scf.h5 does not exist in this case. 

But then it sends also 
 mpirun -np 1 convert4qmc -prefix c4q -orbitals scf.h5 
 
 Obviously scf.h5 does not exist as well. 
 
 After fixing these by hand to push the simulation, unfortunately it still fails  at the vmc with:
     for n,(h5file,kp) in enumerate(zip(result.orb_files,result.kpoints)):
                                       ^^^^^^^^^^^^^^^^
AttributeError: 'obj' object has no attribute 'orb_files'


I Think it would be faster for you to track the error.

**Comment by jtkrogel** [comment]
After discussion with Anoaur, I've confirmed that Nexus is now fully functional in this PR.  Anouar plans to push another fix to PyscfToQmcpack.py.

I will fix the unit test failure issue (which doesn't reproduce on my laptop) in a week or so after this PR goes in and I am back from vacation.

**Comment by jtkrogel** [comment]
If anyone wants to look at the test failure in the interim, it's likely that this will fix the issue (adds a 1e-8 absolute tolerance in addition to the default 1e-6 relative tolerance):
```
assert(text_eq(text,ref_text,atol=1e-8))
```

**Comment by prckent** [comment]
Didn't know about that tolerance option. Will try it. If updating the tolerance that way works, we can probably merge this PR. The default relative tolerance is likely unreliable with zero as one of the values.

We can't merge something that is known to break a deterministic test (recommended installation test 'ctest -L deterministic') and the CI without putting in a bypass (expected failure). This in turn would forgive other errors, so while PRs don't have to have a complete feature (and probably shouldn't if a large feature), they do need to pass in CI somehow.

**Comment by ye-luo** [comment]
I remember some discussion long time ago that twistnum is error-prone and explicit twist coordinates are preferred. Can nexus move away from twistnum?

**Comment by prckent** [comment]
> I remember some discussion long time ago that twistnum is error-prone and explicit twist coordinates are preferred. Can nexus move away from twistnum?

**Review by prckent — APPROVED** [review]
With the tolerance change, CI passes, so we can merge this. There are still homework items to work on in later PRs, but these look to be tied to the converter not Nexus.

Summarizing the current status:

* ctest -L deterministic passes . Core ntest_nexus* tests are good.
* nexus/examples/qmcpack/rsqmc_pyscf/01_h2o_hf_qmc works (was untouched)
* nexus/examples/qmcpack/rsqmc_pyscf/02_diamond_hf_qmc/diamond_pp_hf_gamma.py  seems OK. Got through multiple rounds of wavefunction optimization before I halted it.
* nexus/examples/qmcpack/rsqmc_pyscf/02_diamond_hf_qmc/diamond_pp_hf_twistavg_prim.py aborts at c4q step. Error is not caught but c4q.err has `ERROR HDF5 read failure in hdf_archive::read numAO` (???)
* nexus/examples/qmcpack/rsqmc_pyscf/02_diamond_hf_qmc/diamond_pp_hf_twistavg.py  Completed the conversion, generating 3 scf.twistnum_00n.h5 files and successfully started VMC. Interrupted at this point.

I can provide exact version numbers of all libraries used should that be helpful.

**Pull request merged**